### PR TITLE
Improve visibility of sliceviewer and colorfill zoombox

### LINKED ISSUE #29264: Sliceviewer zoom box is almost impossible to see
image

### Expected behavior
You should be able to see the box you are selecting

### Actual behavior
At the moment the border is 1px dashed black.

With the default colour map that means for 0 data (very common) it is against a dark blue background and very hard to see.

### Suggestions
Thicken the line a bit and make it a contrasting colour to the colour map (certainly the bottom of the colour map)

### PR DESCRIPTION
**Pull request by StephenSmith25**
**Description of work.**
This PR improves the visibility of the sliceviewer and colorfil zoombox. This is achieved by subclassing both the FigureCanvas and AxesImage. The FigureCanvas is subclassed so we can control the pen color and thickness used to draw the annotations from the toolbar. 

The AxesImage is subclassed to add a new method to find the intensity of the image in grayscale. Within this method the intensity varies between 0 and 255, where 0 indicates black (dark image) and 255 white (a light image). The value of grayscale yields the pen color (subject to a threshold), which will be used to draw the zoombox. This calculation is updated every time the image is drawn, e.g if we zoom on a high counting region the pen may change from white to black (depending on the color map).

**To test:**

-	Code review
-	Load MAR27284.nxs
-	Open sliceviewer
-	Test the magnification tool, it will be white to start with as the image is “dark”
-	Zoom in on the high counting data, as you go further in the tool will change to black (as yellow fills the screen)
-	Go back to the home view
-	Click reverse
-	The tool should now be black

Fixes #29264  

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: qt/scientific_interfaces/CMakeLists.txt
Comment (DanielMurphy22) [inline qt/scientific_interfaces/CMakeLists.txt:68]: How come these have been removed, is this related to another PR or a mistake?

File: Framework/PythonInterface/mantid/plots/mantidimage.py
Comment (DanielMurphy22) [inline Framework/PythonInterface/mantid/plots/mantidimage.py:88]: Maybe worth adding a comment here as to where you got these numbers to multiply r, g and b by to give the greyscale intensity? Also is the Max value of grey 255? If so then should the threshold be 100 or 127.5??

**Review by DanielMurphy22 — CHANGES_REQUESTED** [review]
This works really well and the zoombox is always a good colour now as it was only not visible when the main colour of the image was very dark or very bright.

As well as the other comments I've left, I also have a slight edge case to ask about. **I could be tempted** into saying this isn't worth fixing but if you think it would be quick to cover then great.

- If a lot of the data is outside of the colorbar limits then the main color of the image is white. Despite this I am getting white as the zoom tool color.

To reproduce run this and try to zoom:

```
import matplotlib.pyplot as plt
import numpy as np
from matplotlib.colors import LogNorm
from matplotlib.ticker import LogLocator
from mantid.api import AnalysisDataService as ADS

CNCS_7860_event = ADS.retrieve('CNCS_7860_event')

fig, axes = plt.subplots(dpi=258.0, edgecolor='#ffffff', figsize=[8.0, 7.0], num='CNCS_7860_event-1', subplot_kw={'projection': 'mantid'})
cfill = axes.imshow(CNCS_7860_event, aspect='auto', cmap='viridis', distribution=False, label='_image0', origin='lower')
cfill.set_norm(LogNorm(vmin=0.01, vmax=0.14999162156175516))
# If no ticks appear on the color bar remove the subs argument inside the LogLocator below
cbar = fig.colorbar(cfill, ax=[axes], ticks=LogLocator(subs=np.arange(1, 10)), pad=0.06)
cbar.set_label('Counts ($\\mu s$)$^{-1}$')
axes.set_title('CNCS_7860_event')
axes.set_xlabel('Time-of-flight ($\\mu s$)')
axes.set_ylabel('Spectrum')
axes.set_xlim([44163.0, 60830.0])
axes.set_ylim([0.5, 51200.0])

plt.show()
```

File: qt/scientific_interfaces/CMakeLists.txt
Comment (StephenSmith25) [reply qt/scientific_interfaces/CMakeLists.txt:68]: Nope these this is intentional, these tests were still running in qt4 mode, which didn’t support the qt5 backend used for the mantidcanvas class defined here.

**Comment by StephenSmith25** [comment]
> If a lot of the data is outside of the colorbar limits then the main color of the image is white. Despite this I am getting white as the zoom tool color.

This is an issue that’s related to choosing a log scale, as its masking so much of the data. This automatic switching is based on the color of the data pixels, rather than the canvas pixels. In this example we've masked most of the data, therefore we don't have many data pixels and the ones we do have indicate that the image is dark. I think this would fall out of the normal use case as in general we'd expect that the canvas pixels ≈ data pixels. 

In order to base the color off the canvas pixels I would have to render the figure without the axes, titles and legend and then calculate the intensity of the image, which felt like more work (computationally). But would be open to doing it if necessary.

File: Framework/PythonInterface/mantid/plots/mantidimage.py
Comment (StephenSmith25) [reply Framework/PythonInterface/mantid/plots/mantidimage.py:88]: I've added a comment where they were taken from. Yeah the max value would be 255, so half would be 127.5, but that doesn't necessarily mean its halfway between light and dark. Its fairly qualitative in terms of what constitutes "dark" and "light", to me at 127 a few too many images were being classified as dark, 100 felt like a good number - but I guess others would have different interpretations.

image

**Review by DanielMurphy22 — DISMISSED** [review]
Well done, works very well. Thanks for clarifying the choice of threshold. Also, who doesn't like a good reference to Wikipedia!? Disabling the tests on QT4 should be fine as they are also run on QT5. 

Nudge me to re-approve after fixing the merge conflict!

**Review by DanielMurphy22 — DISMISSED** [review]

**Comment by gemmaguest** [comment]
@StephenSmith25 Sorry, this has conflicts again

**Review by DanielMurphy22 — APPROVED** [review]

**Pull request merged**
